zero/src/prover.rs

```diff
@@ -131,29 +135,143 @@ impl BlockProverInput {
             save_inputs_on_error,
         };
 
-        // Segment the batches, prove segments and aggregate them to resulting batch
-        // proofs.
-        let batch_proof_futs: FuturesUnordered<_> = block_generation_inputs
-            .iter()
-            .enumerate()
-            .map(|(idx, txn_batch)| {
-                let segment_data_iterator =
-                    SegmentDataIterator::<Field>::new(txn_batch, Some(max_cpu_len_log));
-
-                Directive::map(IndexedStream::from(segment_data_iterator), &seg_prove_ops)
-                    .fold(&seg_agg_ops)
-                    .run(&proof_runtime.heavy_proof)
-                    .map(move |e| {
-                        e.map(|p| (idx, crate::proof_types::BatchAggregatableProof::from(p)))
-                    })
+        // Generate channels to communicate segments of each batch to a batch proving
+        // task. We generate a segment and send it to the proving task, then
+        // wait for it to be proved before we send another segment. This is done
+        // to avoid memory exhaustion.
```

Comment:
If I understand correctly, this is just to limit the amount of batch per channel to 1 (and then we split them off below to parallelize segment proving).
I'm not sure we'd get memory exhaustion here with a greater buffer length, regarding the size of the data involved, but I don't think it'd bring anything either, so non blocking.

Reply:
We have a proving task and independent channel for every batch in parallel. We generate one by one segments proven in that batch (so channel is for segments of one batch) and we prove them sequentially. On the second though, this could be choking point so I should parametrize this parameter.

Reply:
Changed logic, we consume segments and span parallel batch segment proving tasks without limitation.